# Pandas
Pandas is a popular Python package for data science, and with good reason: it offers powerful, expressive and flexible data structures that make data manipulation and analysis easy, among many other things. 

Pandas deals with the following data structures −

* Series
* DataFrame



#### Series
Series is a one-dimensional array like structure with homogeneous data. For example, the following series is a collection of integers 10, 23, 56, …

In [73]:
my_series = pd.Series([10,23,56,78])
print(my_series)

0    10
1    23
2    56
3    78
dtype: int64


In [74]:
print(type(my_series))

<class 'pandas.core.series.Series'>


It can have any data structure like integer, float, and string. It is useful when you want to perform computation or return a one-dimensional array. A series, by definition, cannot have multiple columns.

You can add the index with index. It helps to name the rows. The length should be equal to the size of the column.

In [79]:
my_series = pd.Series([1., 2., 3.], index=['a', 'b', 'c'])
print(my_series)

a    1.0
b    2.0
c    3.0
dtype: float64


### Pandas Data Frames

They are defined as two-dimensional labeled data structures with columns of potentially different types.

The Pandas DataFrame consists of three main components: the data, the index, and the columns.

The Pandas library is usually imported under the alias pd.


In [56]:
import pandas as pd

#### Create Data frame
You can convert a numpy array to a pandas data frame with pd.DataFrame(). The opposite is also possible. To convert a pandas Data Frame to an array, you can use np.array()


In [80]:
## Numpy to pandas
h = [[1,2],[3,4]] 
df_h = pd.DataFrame(h)
print('Data Frame:', df_h)

Data Frame:    0  1
0  1  2
1  3  4


In [81]:
## Pandas to numpy
df_h_n = np.array(df_h)
print('Numpy array:', df_h_n)

Numpy array: [[1 2]
 [3 4]]


You can also use a dictionary to create a Pandas dataframe.

In [82]:
dic = {'Name': ["John", "Smith"], 'Age': [30, 40]}
pd.DataFrame(data=dic)

Name  Age
0   John   30
1  Smith   40

### Importing Data with read_csv()
The first step to any data science project is to import your data. Often, you'll work with data in Comma Separated Value (CSV) files and run into problems at the very start of your workflow. 



Before you can use pandas to import your data, you need to know where your data is in your filesystem and what your current working directory is.

# OS library
Python allows the developer to use several OS-dependent functionalities with the Python module <b>os</b>. This package abstracts the functionalities of the platform and provides the python functions to navigate, create, delete and modify files and folders.


In [57]:
import os

#### getcwd()
Now, using the getcwd method, we can retrieve the path of the current working directory.

In [58]:
print(os.getcwd())

C:\Users\ab275\Downloads\edureka_freelance\gaurav ipynb\python1


Let's list the folders/files in the current directory using listdir:

In [59]:
print(os.listdir())

['.ipynb_checkpoints', '01_initial_notebook_screen.cb2ea87d9679.png', 'Basics of Python Part 1.ipynb', 'Basics of Python Part 2.ipynb', 'Basics of Python Part 3.ipynb', 'Basics of Python Part 4.ipynb', 'calculation.py', 'continue-statement-flowchart.jpg', 'flowchart-break-statement.jpg', 'forLoop.jpg', 'hello.png', 'numpy_tut1.png', 'osx-install-destination.png', 'osx-install-success.png', 'osx-install-type.png', 'reduce_diagram.png', 'test.txt', 'whileLoopFlowchart.jpg', 'win-install-complete.png']


Let's change the working directory and enter into the directory of data.

In [61]:
os.chdir('Data/')

#### Loading your data
Now that you know what your current working directory is and where the dataset is in your filesystem, you can specify the file path to it. You're now ready to import the CSV file into Python using read_csv() from pandas:


In [83]:
credit_data = pd.read_csv("Credit_Pay.csv")

In the code above, the file path is the main argument to read_csv() and it was specified as the relative file path. The read_csv() function is smart enough to decipher whether it's working with full or relative file paths and convert your flat file as a DataFrame without a problem.

# Inspecting data

* Load the data
* Using head function - to print first few rows of the dataframe
* Using tail function - to print last few rows of the dataframe
* Using describe function - to get the counts, mean, std, min, max and percentile of the dataframe
* Using info function - to see the datatypes of the columns of the dataframe

Data has been load already using read.csv() in the object "credit_data"

In [85]:
print(credit_data.head())

   ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0   1    20000.0  Female  Post Graduate  Married   24              2   
1   2   120000.0  Female  Post Graduate   Single   26             -1   
2   3    90000.0  Female  Post Graduate   Single   34              0   
3   4    50000.0  Female  Post Graduate  Married   37              0   
4   5    50000.0    Male  Post Graduate  Married   57             -1   

   REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_MAY  \
0              2             -1             -1  ...         689.0   
1              2              0              0  ...        2682.0   
2              0              0              0  ...       13559.0   
3              0              0              0  ...       49291.0   
4              0             -1              0  ...       35835.0   

   BILL_AMT_APR  BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  \
0           0.0           0.0           0.0          0.0        689.0   
1     

In [86]:
print(credit_data.tail())

          ID  LIMIT_BAL   SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
29995  29996   220000.0  Male    High School  Married   39              0   
29996  29997   150000.0  Male    High School   Single   43             -1   
29997  29998    30000.0  Male  Post Graduate   Single   37              4   
29998  29999    80000.0  Male    High School  Married   41              1   
29999  30000    50000.0  Male  Post Graduate  Married   46              0   

       REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_MAY  \
29995              0              0              0  ...      208365.0   
29996             -1             -1             -1  ...        3502.0   
29997              3              2             -1  ...        2758.0   
29998             -1              0              0  ...       76304.0   
29999              0              0              0  ...       49764.0   

       BILL_AMT_APR  BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  \
29995       88004.0  

In [87]:
print(credit_data.describe())

                 ID       LIMIT_BAL           AGE  REPAYMENT_JUL  \
count  30000.000000    30000.000000  30000.000000   30000.000000   
mean   15000.500000   167484.322667     35.485500      -0.016700   
std     8660.398374   129747.661567      9.217904       1.123802   
min        1.000000    10000.000000     21.000000      -2.000000   
25%     7500.750000    50000.000000     28.000000      -1.000000   
50%    15000.500000   140000.000000     34.000000       0.000000   
75%    22500.250000   240000.000000     41.000000       0.000000   
max    30000.000000  1000000.000000     79.000000       8.000000   

       REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  REPAYMENT_MAR  \
count   30000.000000   30000.000000   30000.000000   30000.000000   
mean       -0.133767      -0.166200      -0.220667      -0.266200   
std         1.197186       1.196868       1.169139       1.133187   
min        -2.000000      -2.000000      -2.000000      -2.000000   
25%        -1.000000      -1.000000      -

In [88]:
print(credit_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 24 columns):
ID               30000 non-null int64
LIMIT_BAL        30000 non-null float64
SEX              30000 non-null object
EDUCATION        30000 non-null object
MARRIAGE         30000 non-null object
AGE              30000 non-null int64
REPAYMENT_JUL    30000 non-null int64
REPAYMENT_JUN    30000 non-null int64
REPAYMENT_MAY    30000 non-null int64
REPAYMENT_APR    30000 non-null int64
REPAYMENT_MAR    30000 non-null int64
REPAYMENT_FEB    30000 non-null int64
BILL_AMT_JUL     30000 non-null float64
BILL_AMT_JUN     30000 non-null float64
BILL_AMT_MAY     30000 non-null float64
BILL_AMT_APR     30000 non-null float64
BILL_AMT_MAR     30000 non-null float64
BILL_AMT_FEB     30000 non-null float64
PAY_AMT_JUL      30000 non-null float64
PAY_AMT_JUN      30000 non-null float64
PAY_AMT_MAY      30000 non-null float64
PAY_AMT_APR      30000 non-null float64
PAY_AMT_MAR      30000 non-nul

## Shape
shape method is used to check the number of rows and columns of dataframe. First number denotes the number of rows and second number denotes the number of columns


In [151]:
print(credit_data.shape)

(30000, 25)


## Check for Missing Values

In [149]:
# To see if any column has NaN or Null values
print(credit_data.isnull().sum())

ID               0
LIMIT_BAL        0
SEX              0
EDUCATION        0
MARRIAGE         0
AGE              0
REPAYMENT_JUL    0
REPAYMENT_JUN    0
REPAYMENT_MAY    0
REPAYMENT_APR    0
REPAYMENT_MAR    0
REPAYMENT_FEB    0
BILL_AMT_JUL     0
BILL_AMT_JUN     0
BILL_AMT_MAY     0
BILL_AMT_APR     0
BILL_AMT_MAR     0
BILL_AMT_FEB     0
PAY_AMT_JUL      0
PAY_AMT_JUN      0
PAY_AMT_MAY      0
PAY_AMT_APR      0
PAY_AMT_MAR      0
PAY_AMT_FEB      0
MARRIED          0
dtype: int64


WooHoo! Our data is completely free of missing values.

#### Slice data
You can use the column name to extract data in a particular column.

In [91]:
### Using name
credit_data['EDUCATION']

0        Post Graduate
1        Post Graduate
2        Post Graduate
3        Post Graduate
4        Post Graduate
             ...      
29995      High School
29996      High School
29997    Post Graduate
29998      High School
29999    Post Graduate
Name: EDUCATION, Length: 30000, dtype: object

To select multiple columns, you need to use two times the bracket, [[..,..]]

The first pair of bracket means you want to select columns, the second pairs of bracket tells what columns you want to return.

In [92]:
print(credit_data[['EDUCATION','MARRIAGE']])

           EDUCATION MARRIAGE
0      Post Graduate  Married
1      Post Graduate   Single
2      Post Graduate   Single
3      Post Graduate  Married
4      Post Graduate  Married
...              ...      ...
29995    High School  Married
29996    High School   Single
29997  Post Graduate   Single
29998    High School  Married
29999  Post Graduate  Married

[30000 rows x 2 columns]


The code below returns the first three rows.

The first pair of bracket means you want to select columns, the second pairs of bracket tells what columns you want to return.

You can slice the rows with :

In [93]:
### using a slice for row
print(credit_data[0:3])

   ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0   1    20000.0  Female  Post Graduate  Married   24              2   
1   2   120000.0  Female  Post Graduate   Single   26             -1   
2   3    90000.0  Female  Post Graduate   Single   34              0   

   REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_MAY  \
0              2             -1             -1  ...         689.0   
1              2              0              0  ...        2682.0   
2              0              0              0  ...       13559.0   

   BILL_AMT_APR  BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  \
0           0.0           0.0           0.0          0.0        689.0   
1        3272.0        3455.0        3261.0          0.0       1000.0   
2       14331.0       14948.0       15549.0       1518.0       1500.0   

   PAY_AMT_MAY  PAY_AMT_APR  PAY_AMT_MAR  PAY_AMT_FEB  
0          0.0          0.0          0.0          0.0  
1       1000.0       1000.0  

The code above returns the first three rows.

## loc
The loc function is used to select columns by names. As usual, the values before the coma stand for the rows and after refer to the column. You need to use the brackets to select more than one column.

In [95]:
## Multi col
print(credit_data.loc[:,['EDUCATION','MARRIAGE']])


           EDUCATION MARRIAGE
0      Post Graduate  Married
1      Post Graduate   Single
2      Post Graduate   Single
3      Post Graduate  Married
4      Post Graduate  Married
...              ...      ...
29995    High School  Married
29996    High School   Single
29997  Post Graduate   Single
29998    High School  Married
29999  Post Graduate  Married

[30000 rows x 2 columns]


## iloc
There is another method to select multiple rows and columns in Pandas. You can use iloc[]. This method uses the index instead of the columns name. The code below returns the same data frame as above.

In [97]:
## Multi col
print(credit_data.iloc[:,3:5])

           EDUCATION MARRIAGE
0      Post Graduate  Married
1      Post Graduate   Single
2      Post Graduate   Single
3      Post Graduate  Married
4      Post Graduate  Married
...              ...      ...
29995    High School  Married
29996    High School   Single
29997  Post Graduate   Single
29998    High School  Married
29999  Post Graduate  Married

[30000 rows x 2 columns]


In [99]:
# Lets create a variable using loc method
# Create a variable 'MARRIED' where 1 denotes married and 0 denotes not married
credit_data['MARRIED'] = 0
# Here we first create a variable within the dataframe with the name 'MARRIED' and all values initially assigned to 0.
credit_data.loc[credit_data['MARRIAGE']=='Married','MARRIED'] = 1
# Now, using loc method , we are subsetting the dataset to select only those rows which has 'MARRIAGE' variable value 'Married' 
# and assigning the values of 'MARRIAGE' as 1

In [101]:
print(credit_data[['MARRIAGE','MARRIED']].head())

  MARRIAGE  MARRIED
0  Married        1
1   Single        0
2   Single        0
3  Married        1
4  Married        1


#### Drop a column
You can drop columns using pd.drop()

In [102]:
print(credit_data.drop(columns=['MARRIED']))

          ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0          1    20000.0  Female  Post Graduate  Married   24              2   
1          2   120000.0  Female  Post Graduate   Single   26             -1   
2          3    90000.0  Female  Post Graduate   Single   34              0   
3          4    50000.0  Female  Post Graduate  Married   37              0   
4          5    50000.0    Male  Post Graduate  Married   57             -1   
...      ...        ...     ...            ...      ...  ...            ...   
29995  29996   220000.0    Male    High School  Married   39              0   
29996  29997   150000.0    Male    High School   Single   43             -1   
29997  29998    30000.0    Male  Post Graduate   Single   37              4   
29998  29999    80000.0    Male    High School  Married   41              1   
29999  30000    50000.0    Male  Post Graduate  Married   46              0   

       REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR 

Since we did not assign dropped dataframe to any variable, the actual dataframe still has 'MARRIED' variable

In [103]:
print(credit_data.head())

   ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0   1    20000.0  Female  Post Graduate  Married   24              2   
1   2   120000.0  Female  Post Graduate   Single   26             -1   
2   3    90000.0  Female  Post Graduate   Single   34              0   
3   4    50000.0  Female  Post Graduate  Married   37              0   
4   5    50000.0    Male  Post Graduate  Married   57             -1   

   REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_APR  \
0              2             -1             -1  ...           0.0   
1              2              0              0  ...        3272.0   
2              0              0              0  ...       14331.0   
3              0              0              0  ...       28314.0   
4              0             -1              0  ...       20940.0   

   BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  PAY_AMT_MAY  \
0           0.0           0.0          0.0        689.0          0.0   
1       

print(credit_data.head())

## Concatenation
You can concatenate two DataFrame in Pandas. You can use pd.concat()

First of all, you need to create two DataFrames. So far so good, you are already familiar with dataframe creation.

Let's create two dataframes from our credit_data.
Divide the dataframe into two dataframes : Married_data, Unmarried_data

In [109]:
credit_married_data = credit_data[credit_data['MARRIED']==1]
print(credit_married_data.head())

    ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0    1    20000.0  Female  Post Graduate  Married   24              2   
3    4    50000.0  Female  Post Graduate  Married   37              0   
4    5    50000.0    Male  Post Graduate  Married   57             -1   
8    9   140000.0  Female    High School  Married   28              0   
17  18   320000.0    Male       Graduate  Married   49              0   

    REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_APR  \
0               2             -1             -1  ...           0.0   
3               0              0              0  ...       28314.0   
4               0             -1              0  ...       20940.0   
8               0              2              0  ...       12211.0   
17              0              0             -1  ...       70074.0   

    BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  PAY_AMT_MAY  \
0            0.0           0.0          0.0        689.0          0

In [106]:
credit_unmarried_data = credit_data[credit_data['MARRIED']==0]
print(credit_unmarried_data.head())

   ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
1   2   120000.0  Female  Post Graduate   Single   26             -1   
2   3    90000.0  Female  Post Graduate   Single   34              0   
5   6    50000.0    Male       Graduate   Single   37              0   
6   7   500000.0    Male       Graduate   Single   29              0   
7   8   100000.0  Female  Post Graduate   Single   23              0   

   REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_APR  \
1              2              0              0  ...        3272.0   
2              0              0              0  ...       14331.0   
5              0              0              0  ...       19394.0   
6              0              0              0  ...      542653.0   
7             -1             -1              0  ...         221.0   

   BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  PAY_AMT_MAY  \
1        3455.0        3261.0          0.0       1000.0       1000.0   
2       

Now, concatenate both dataframes to combine into one dataframe.

In [111]:
credit_concat = pd.concat([credit_married_data,credit_unmarried_data]) 
print(credit_concat.head())

    ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0    1    20000.0  Female  Post Graduate  Married   24              2   
3    4    50000.0  Female  Post Graduate  Married   37              0   
4    5    50000.0    Male  Post Graduate  Married   57             -1   
8    9   140000.0  Female    High School  Married   28              0   
17  18   320000.0    Male       Graduate  Married   49              0   

    REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_APR  \
0               2             -1             -1  ...           0.0   
3               0              0              0  ...       28314.0   
4               0             -1              0  ...       20940.0   
8               0              2              0  ...       12211.0   
17              0              0             -1  ...       70074.0   

    BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  PAY_AMT_MAY  \
0            0.0           0.0          0.0        689.0          0

Since, unmarried data is concatenated after married data, all unmarried rows will come after all married rows. To check, use tail().

In [112]:
print(credit_concat.tail())

          ID  LIMIT_BAL   SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
29989  29990   150000.0  Male       Graduate   Single   35             -1   
29993  29994   100000.0  Male       Graduate   Single   38              0   
29994  29995    80000.0  Male  Post Graduate   Single   34              2   
29996  29997   150000.0  Male    High School   Single   43             -1   
29997  29998    30000.0  Male  Post Graduate   Single   37              4   

       REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_APR  \
29989             -1             -1             -1  ...         780.0   
29993             -1             -1              0  ...       70626.0   
29994              2              2              2  ...       77519.0   
29996             -1             -1             -1  ...        8979.0   
29997              3              2             -1  ...       20878.0   

       BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  PAY_AMT_MAY  \
29989           0.0   

Let's create a small subset of data from the original data by taking top 20 rows and three columns.

In [119]:
credit_small = credit_data.iloc[:20,2:5]
print(credit_small.shape)

(20, 3)


## Drop_duplicates
If a dataset can contain duplicates information use, `drop_duplicates` is an easy to exclude duplicate rows.

In [120]:
print(credit_small)

       SEX      EDUCATION MARRIAGE
0   Female  Post Graduate  Married
1   Female  Post Graduate   Single
2   Female  Post Graduate   Single
3   Female  Post Graduate  Married
4     Male  Post Graduate  Married
5     Male       Graduate   Single
6     Male       Graduate   Single
7   Female  Post Graduate   Single
8   Female    High School  Married
9     Male    High School   Single
10  Female    High School   Single
11  Female       Graduate   Single
12  Female  Post Graduate   Single
13    Male  Post Graduate   Single
14    Male       Graduate   Single
15  Female    High School   Others
16    Male       Graduate   Single
17    Male       Graduate  Married
18  Female       Graduate  Married
19  Female       Graduate   Single


We can see that `credit_small` has many duplicate observations. Let's remove those duplicates from the dataset.

In [127]:
credit_small_no_dup = credit_small.drop_duplicates()
print(credit_small_no_dup)

       SEX      EDUCATION MARRIAGE
0   Female  Post Graduate  Married
1   Female  Post Graduate   Single
4     Male  Post Graduate  Married
5     Male       Graduate   Single
8   Female    High School  Married
9     Male    High School   Single
10  Female    High School   Single
11  Female       Graduate   Single
13    Male  Post Graduate   Single
15  Female    High School   Others
17    Male       Graduate  Married
18  Female       Graduate  Married


In [126]:
print(credit_small_no_dup.shape)

(12, 3)


We can see that now it has only 12 rows, i.e., 8 rows which were duplicates have been dropped.

Let's say we wanted to drop duplicates only basis either a single column or some column subset. 

In [128]:
credit_small_no_dup2 = credit_small.drop_duplicates(['EDUCATION','MARRIAGE'])
print(credit_small_no_dup2)

       SEX      EDUCATION MARRIAGE
0   Female  Post Graduate  Married
1   Female  Post Graduate   Single
5     Male       Graduate   Single
8   Female    High School  Married
9     Male    High School   Single
15  Female    High School   Others
17    Male       Graduate  Married


In [129]:
print(credit_small_no_dup2.shape)

(7, 3)


Now, it has only 7 rows.

## Sort values
You can sort value with sort_values(). You need to specify basis which column you want to sort your dataframe.

In [131]:
print(credit_small.sort_values('SEX'))

       SEX      EDUCATION MARRIAGE
0   Female  Post Graduate  Married
15  Female    High School   Others
12  Female  Post Graduate   Single
11  Female       Graduate   Single
10  Female    High School   Single
18  Female       Graduate  Married
8   Female    High School  Married
19  Female       Graduate   Single
3   Female  Post Graduate  Married
2   Female  Post Graduate   Single
1   Female  Post Graduate   Single
7   Female  Post Graduate   Single
6     Male       Graduate   Single
5     Male       Graduate   Single
4     Male  Post Graduate  Married
13    Male  Post Graduate   Single
14    Male       Graduate   Single
16    Male       Graduate   Single
17    Male       Graduate  Married
9     Male    High School   Single


In [132]:
print(credit_small.sort_values(['SEX','EDUCATION']))

       SEX      EDUCATION MARRIAGE
11  Female       Graduate   Single
18  Female       Graduate  Married
19  Female       Graduate   Single
8   Female    High School  Married
10  Female    High School   Single
15  Female    High School   Others
0   Female  Post Graduate  Married
1   Female  Post Graduate   Single
2   Female  Post Graduate   Single
3   Female  Post Graduate  Married
7   Female  Post Graduate   Single
12  Female  Post Graduate   Single
5     Male       Graduate   Single
6     Male       Graduate   Single
14    Male       Graduate   Single
16    Male       Graduate   Single
17    Male       Graduate  Married
9     Male    High School   Single
4     Male  Post Graduate  Married
13    Male  Post Graduate   Single


## Rename
You can use rename to rename a column in Pandas. The first value is the current column name and the second value is the new column name.

In [134]:
print(credit_small.rename(columns={"SEX": "sex_of_customer", "EDUCATION": "edu_of_customer"}))

   sex_of_customer edu_of_customer MARRIAGE
0           Female   Post Graduate  Married
1           Female   Post Graduate   Single
2           Female   Post Graduate   Single
3           Female   Post Graduate  Married
4             Male   Post Graduate  Married
5             Male        Graduate   Single
6             Male        Graduate   Single
7           Female   Post Graduate   Single
8           Female     High School  Married
9             Male     High School   Single
10          Female     High School   Single
11          Female        Graduate   Single
12          Female   Post Graduate   Single
13            Male   Post Graduate   Single
14            Male        Graduate   Single
15          Female     High School   Others
16            Male        Graduate   Single
17            Male        Graduate  Married
18          Female        Graduate  Married
19          Female        Graduate   Single


Find the mean LIMIT_BAL of male customers who are married and post graduates.

In [143]:
# For this, first we create the data and then find the mean.
filtered_data1 = credit_data[(credit_data['SEX']=='Male') & (credit_data['MARRIAGE']=='Married') & (credit_data['EDUCATION']=='Post Graduate')]

In [145]:
print(filtered_data1.shape)
print(filtered_data1.LIMIT_BAL.mean())

(2370, 25)
162101.2658227848


Rs 162101 is the mean value of Limit_Bal for male customers who are married and post graduates.

Find the max Age of Female customers in one line of code.

In [147]:
print(credit_data.loc[credit_data['SEX']=='Female','AGE'].max())

75


Repeat above code for male customers.

In [148]:
print(credit_data.loc[credit_data['SEX']=='Male','AGE'].max())

79


Groupby
An easy way to see the data is to use the groupby method. This method can help you to summarize the data by group. Below is a list of methods available with groupby:
* count: count
* min: min
* max: max
* mean: mean
* median: median
* standard deviation: std


Inside groupby(), you can use the column you want to apply the method.

Let's have a look at a single grouping with the credit_data dataset. Find the mean of Limit_Bal of all the customers by type of Education.

In [155]:
print(credit_data.groupby(['EDUCATION'])['LIMIT_BAL'].mean())

EDUCATION
Graduate         212956.069910
High School      126550.270490
Others           220894.308943
Post Graduate    147062.437634
Unknown          167205.797101
Name: LIMIT_BAL, dtype: float64


You can also group by multiple columns. 

For instance, find the minimum of age by type of EDUCATION AND SEX

In [158]:
print(credit_data.groupby(['SEX','EDUCATION'])['AGE'].max())

SEX     EDUCATION    
Female  Graduate         70
        High School      74
        Others           59
        Post Graduate    75
        Unknown          64
Male    Graduate         79
        High School      73
        Others           52
        Post Graduate    75
        Unknown          66
Name: AGE, dtype: int64


Find the mean of all numeric columns in the dataset.

In [160]:
print(credit_data.mean(axis='rows'))

ID                15000.500000
LIMIT_BAL        167484.322667
AGE                  35.485500
REPAYMENT_JUL        -0.016700
REPAYMENT_JUN        -0.133767
REPAYMENT_MAY        -0.166200
REPAYMENT_APR        -0.220667
REPAYMENT_MAR        -0.266200
REPAYMENT_FEB        -0.291100
BILL_AMT_JUL      51223.330900
BILL_AMT_JUN      49179.075167
BILL_AMT_MAY      47013.154800
BILL_AMT_APR      43262.948967
BILL_AMT_MAR      40311.400967
BILL_AMT_FEB      38871.760400
PAY_AMT_JUL        5663.580500
PAY_AMT_JUN        5921.163500
PAY_AMT_MAY        5225.681500
PAY_AMT_APR        4826.076867
PAY_AMT_MAR        4799.387633
PAY_AMT_FEB        5215.502567
MARRIED               0.455300
dtype: float64


## isin()
Pandas isin() method is used to filter data frames. isin() method helps in selecting rows with having a particular(or Multiple) value in a particular column.

Create a subset of the data to get details of customers who are Graudated and Post Graduated

In [163]:
credit_data_subset = credit_data[credit_data['EDUCATION'].isin(['Graduate','Post Graduate'])]
print(credit_data_subset.head())

   ID  LIMIT_BAL     SEX      EDUCATION MARRIAGE  AGE  REPAYMENT_JUL  \
0   1    20000.0  Female  Post Graduate  Married   24              2   
1   2   120000.0  Female  Post Graduate   Single   26             -1   
2   3    90000.0  Female  Post Graduate   Single   34              0   
3   4    50000.0  Female  Post Graduate  Married   37              0   
4   5    50000.0    Male  Post Graduate  Married   57             -1   

   REPAYMENT_JUN  REPAYMENT_MAY  REPAYMENT_APR  ...  BILL_AMT_APR  \
0              2             -1             -1  ...           0.0   
1              2              0              0  ...        3272.0   
2              0              0              0  ...       14331.0   
3              0              0              0  ...       28314.0   
4              0             -1              0  ...       20940.0   

   BILL_AMT_MAR  BILL_AMT_FEB  PAY_AMT_JUL  PAY_AMT_JUN  PAY_AMT_MAY  \
0           0.0           0.0          0.0        689.0          0.0   
1       

## value_counts()
To count the values of categorical variable

In [164]:
print(credit_data['EDUCATION'].value_counts())

Post Graduate    14030
Graduate         10585
High School       4917
Unknown            345
Others             123
Name: EDUCATION, dtype: int64


## Pivot Table
Create a pivot table of the dataset with EDUCATION AND SEX as row indexes and mean of LIMIT_BAL as column for each row index.

In [167]:
pd.pivot_table(credit_data, values ='LIMIT_BAL', index =['SEX'], columns =['EDUCATION'], aggfunc = np.mean) 

EDUCATION       Graduate    High School         Others  Post Graduate  \
SEX                                                                     
Female     208557.213930  130362.145542  233209.876543  155232.902033   
Male       219251.263206  120943.557789  197142.857143  133902.121325   

EDUCATION        Unknown  
SEX                       
Female     170184.331797  
Male       162156.250000

## Options
pandas has an options system that lets you customize some aspects of its behaviour, display-related options being those the user is most likely to adjust.

Options have a full "dotted-style", case-insensitive name (e.g. display.max_rows). You can get/set options directly as attributes of the top-level options attribute.

Here, we will discuss one major which wille be often used.

While presenting the data, showing the data in the required format is also an important and crucial part. Sometimes, the value is so big that we want to show only desired part of this or we can say in some desired format.

We will make use of pandas' display options for this.

In [1]:
# Round off the column values to two decimal places
import pandas as pd 
  
# create the data dictionary 
data = {'Month' : ['January', 'February', 'March', 'April'], 
     'Expense': [ 21525220.653, 31125840.875, 23135428.768, 56245263.942]} 
  
# create the dataframe 
dataframe = pd.DataFrame(data, columns = ['Month', 'Expense']) 
  
print("Given Dataframe :\n", dataframe) 
  
# round to two decimal places in python pandas 
pd.options.display.float_format = '{:.2f}'.format
  
print('\nResult :\n', dataframe) 

Given Dataframe :
       Month       Expense
0   January  2.152522e+07
1  February  3.112584e+07
2     March  2.313543e+07
3     April  5.624526e+07

Result :
       Month     Expense
0   January 21525220.65
1  February 31125840.88
2     March 23135428.77
3     April 56245263.94


We are importing the library, creating some data, print that data and Expense is in scientific notation.

We have used pandas options display float_format ,method to round to two decimal places.